snuba/query/snql/parser.py

```diff
@@ -108,31 +185,172 @@ class OrTuple(NamedTuple):
     exp: Expression
 
 
+class EntityTuple(NamedTuple):
+    alias: str
+    name: str
+    sample_rate: Optional[float]
+
+
+class RelationshipTuple(NamedTuple):
+    lhs: EntityTuple
+    relationship: str
+    rhs: EntityTuple
+
+
+class JoinGraph(NamedTuple):
```

Comment:
All the code for validating and parsing JOIN queries is here. It seems overly complex to me but I'm not sure how to simplify it down.

Reply:
I think I have an idea about this.
I believe the root cause of the complexity lies in the graph representation that requires you to do the intermediate `_flatten` step. 
What if instead it was represented as a tree? Then creating the join clause would be achievable with a tree traversal (I think any of the traversal types will work). That acts as a reduction. The state you carre over during the traversal is the Join clause. At each step  (which is a node you are traversing) you augment it by creating a new root join between the existing join (on the left side) and the new individual node (on the right side).

Example:

Graph representation. This is actually a set of trees
```
class Edge:
     destination: Node
     conditions: # how you want to represent it

class Node:
    entity: Entity
    children: Set[Edge]

class Trees:
    roots: Set[Node]
    nodes: Set[Node]
    children: Set[Edge]
    
    def add_edge(lhs: Entity, rhs: Entity, conditions[...]
          # retrieve both rhs and lhs from the set of nodes.
          # if only lhs is there, append the edge to that node
          # if only rhs is there, add a new root and make rhs one of its children (it's better be a root otherwise the query is invalid). Remove rhs from the roots, since it is nnon more
          # if none is there, create a new root and add the child
          # if both are threre, rhs must be root. now make rhs a child of lhs and remove it from roots
```

Now if at the end this structure has multiple roots it is an invalid query.

Now you just traverse this tree recursively. At each step you return the current join clause that starts empty

```
def add_to_join(root: Node, join: Optional[JoinClause],condition) -> JoinClause
      # take care of the empty case
     join = Join(join, root, condition)
     for c in children:
         join = add_to_join(join, c, root_c_condition)
         
     return join
     # I think it may be right, then the original tree can be rebuilt by in order traversing this.
```

Of course when building the whole algorithm it will be more complex but I suspect it will still be simpler than the one in this PR since there is no need for the flattening step.

What do you think ?